crates/sui-types/src/programmable_transaction_builder.rs

```diff
@@ -0,0 +1,244 @@
+        package: ObjectID,
+        module: Identifier,
+        function: Identifier,
+        type_arguments: Vec<TypeTag>,
+        call_args: Vec<CallArg>,
+    ) -> anyhow::Result<()> {
+        let mut arguments = vec![];
+        for call_arg in call_args {
+            let arg = match call_arg {
+                call_arg @ (CallArg::Pure(_) | CallArg::Object(_)) => self.input(call_arg),
+                CallArg::ObjVec(objs) if objs.is_empty() => {
+                    anyhow::bail!(
+                        "Empty ObjVec is not supported in programmable transactions \
+                        without a type annotation"
+                    )
+                }
+                CallArg::ObjVec(objs) => {
+                    let make_vec_args = objs.into_iter().map(|obj| self.obj(obj)).collect();
+                    self.command(Command::MakeMoveVec(None, make_vec_args))
+                }
+            };
+            arguments.push(arg)
+        }
+        self.command(Command::MoveCall(Box::new(ProgrammableMoveCall {
+            package,
+            module,
+            function,
+            type_arguments,
+            arguments,
+        })));
+        Ok(())
+    }
+
+    pub fn publish(&mut self, modules: Vec<Vec<u8>>) {
+        self.commands.push(Command::Publish(modules))
+    }
+
+    pub fn transfer_object(&mut self, recipient: SuiAddress, object_ref: ObjectRef) {
+        let rec_arg = self.pure(recipient).unwrap();
+        let obj_arg = self.obj(ObjectArg::ImmOrOwnedObject(object_ref));
+        self.commands
+            .push(Command::TransferObjects(vec![obj_arg], rec_arg));
+    }
+
+    pub fn transfer_sui(&mut self, recipient: SuiAddress, amount: Option<u64>) {
+        let rec_arg = self.pure(recipient).unwrap();
+        let coin_arg = if let Some(amount) = amount {
+            let amt_arg = self.pure(amount).unwrap();
+            self.command(Command::SplitCoin(Argument::GasCoin, amt_arg))
+        } else {
+            Argument::GasCoin
+        };
+        self.command(Command::TransferObjects(vec![coin_arg], rec_arg));
+    }
+
+    pub fn pay_all_sui(&mut self, recipient: SuiAddress) {
+        let rec_arg = self.pure(recipient).unwrap();
+        self.command(Command::TransferObjects(vec![Argument::GasCoin], rec_arg));
+    }
+
+    /// Will fail to generate if recipients and amounts do not have the same lengths
+    pub fn pay_sui(
+        &mut self,
+        recipients: Vec<SuiAddress>,
+        amounts: Vec<u64>,
+    ) -> anyhow::Result<()> {
+        self.pay_impl(recipients, amounts, Argument::GasCoin)
+    }
+
+    /// Will fail to generate if recipients and amounts do not have the same lengths.
+    /// Or if coins is empty
+    pub fn pay(
+        &mut self,
+        coins: Vec<ObjectRef>,
+        recipients: Vec<SuiAddress>,
+        amounts: Vec<u64>,
+    ) -> anyhow::Result<()> {
+        let mut coins = coins.into_iter();
+        let Some(coin) = coins.next()
+        else {
```

Comment:
nit: Funky formatting

Reply:
I think this is the RFC formatting for let else

Reply:
https://github.com/rust-lang/rustfmt/pull/5690

Reply:
I think the RFC formatting for this let-else would be 

```rust
let Some(coin) = coins.next() else {
    anyhow::bail!("coins vector is empty");
};
```

According to [this section](https://github.com/calebcartwright/rust/blob/c694d07c6413ba55caa10b9f8b853df7a7792e7c/src/doc/style-guide/src/statements.md#single-line-let-else-statements)

Reply:
Although I've got to say, the rules as written in the RFC are super ambiguous in general, and I for one welcome our `rustfmt` overloads, whenever they choose to impose their iron will on `let-else`.  Until then, I don't have a strong opposition to this formatting, it just scanned strangely for me.